controller/api/destination/endpoint_translator.go

```diff
@@ -339,19 +343,69 @@ func (et *endpointTranslator) filterAddresses() watcher.AddressSet {
 func (et *endpointTranslator) diffEndpoints(filtered watcher.AddressSet) (watcher.AddressSet, watcher.AddressSet) {
 	add := make(map[watcher.ID]watcher.Address)
 	remove := make(map[watcher.ID]watcher.Address)
+	skipped := make(map[watcher.ID]watcher.Address)
 
+	// Process additions
 	for id, new := range filtered.Addresses {
 		old, ok := et.filteredSnapshot.Addresses[id]
-		if !ok {
-			add[id] = new
-		} else if !reflect.DeepEqual(old, new) {
-			add[id] = new
+		if ok && reflect.DeepEqual(old, new) {
+			continue
+		}
+
+		if id.IPFamily == corev1.IPv6Protocol && !et.preferIPv6 {
+			continue
+		}
+
+		if id.IPFamily == corev1.IPv4Protocol && et.preferIPv6 {
+			// Only consider IPv4 address for which there's not already an IPv6 alternative
+			if altID, _ := et.altFamily(id, corev1.IPv6Protocol); altID != nil {
+				skipped[id] = new
+				continue
+			}
+		}
+
+		add[id] = new
+	}
+
+	// Remove IPv4 alternatives when an IPv6 one appears
+	if et.preferIPv6 {
+		for id := range add {
+			if id.IPFamily != corev1.IPv6Protocol {
+				continue
+			}
+
+			if altID, addr := et.altFamily(id, corev1.IPv4Protocol); altID != nil {
+				if _, ok := skipped[*altID]; !ok {
+					remove[*altID] = *addr
+				}
+			}
 		}
 	}
 
+	// Process removals
 	for id, address := range et.filteredSnapshot.Addresses {
-		if _, ok := filtered.Addresses[id]; !ok {
-			remove[id] = address
+		if _, ok := filtered.Addresses[id]; ok {
+			continue
+		}
+
+		if id.IPFamily == corev1.IPv6Protocol && !et.preferIPv6 {
+			continue
+		}
+
+		if id.IPFamily == corev1.IPv4Protocol && et.preferIPv6 {
+			// Only consider IPv4 address for which there's not already an IPv6 alternative
+			if altID, _ := et.altFamily(id, corev1.IPv6Protocol); altID != nil {
+				continue
+			}
+		}
+
+		remove[id] = address
+
+		if id.IPFamily == corev1.IPv6Protocol && et.preferIPv6 {
```

Comment:
Is this block saying that if the current snapshot contains both IPV6 and IPV4 versions of an address, we should add the IPV4? I don't understand this.

Reply:
It means that if we're removing an IPv6 and we already know about its IPv4 counterpart, send an Add event for the IPv4. This would happen for example when a dual-stack service is updated to be single-stack IPv4; the event for deletion of the IPv6 ES is received and we'd like clients to fall back to the IPv4 ES.